graphql/src/store/prefetch.rs

```diff
@@ -658,17 +602,73 @@ fn execute_selection_set<'a>(
     }
 }
 
+#[derive(Default)]
+struct CollectedResponseKey<'a> {
+    interface_cond: Option<&'a s::InterfaceType>,
+    interface_fields: Vec<&'a q::Field>,
+    concrete_types: IndexMap<ObjectCondition<'a>, Vec<&'a q::Field>>,
+}
+
+impl<'a> CollectedResponseKey<'a> {
+    fn collect_field(&mut self, type_condition: ObjectOrInterface<'a>, field: &'a q::Field) {
+        match type_condition {
+            ObjectOrInterface::Interface(i) => {
+                self.interface_cond = Some(i);
+                self.interface_fields.push(field);
+            }
+            ObjectOrInterface::Object(o) => {
+                self.concrete_types
+                    .entry(ObjectCondition(o))
+                    .or_default()
+                    .push(field);
+            }
+        }
+    }
+}
+
 /// Collects fields of a selection set. The resulting map indicates for each
 /// response key from which types to fetch what fields to express the effect
 /// of fragment spreads
 fn collect_fields<'a>(
     ctx: &'a ExecutionContext<impl Resolver>,
-    object_type: &ObjectOrInterface,
-    selection_sets: Vec<&'a q::SelectionSet>,
-    visited_fragments: Option<HashSet<&'a q::Name>>,
-) -> HashMap<&'a String, HashMap<TypeCondition, Vec<&'a q::Field>>> {
-    let mut visited_fragments = visited_fragments.unwrap_or_default();
-    let mut grouped_fields: HashMap<_, HashMap<_, Vec<_>>> = HashMap::new();
+    parent_ty: ObjectOrInterface<'a>,
+    selection_sets: impl Iterator<Item = &'a q::SelectionSet>,
+) -> Result<IndexMap<&'a String, CollectedResponseKey<'a>>, QueryExecutionError> {
+    let mut grouped_fields = IndexMap::new();
+    collect_fields_inner(
+        ctx,
+        parent_ty,
+        selection_sets,
+        &mut HashSet::new(),
+        &mut grouped_fields,
+    )?;
+
+    // For interfaces, if a response key occurs both under the interface and under concrete types,
+    // we want to add the fields selected at the interface level to the selections in the specific
+    // concrete types, effectively merging the selection sets.
+    // See also: e0d6da3e-60cf-41a5-b83c-b60a7a766d4a
+    for collected_response_key in grouped_fields.values_mut() {
+        for concrete_type_fields in collected_response_key.concrete_types.values_mut() {
```

Comment:
This `for` loop should be a function `merge_interface_fields` or maybe just `finish` on `CollectedResponseKey`

Reply:
I don't think that would gain us much clarity, and that method would only ever be called here, so I'm fine with how this is now.